tbpore/cli.py

```diff
@@ -29,3 +47,214 @@ def handle_parse_result(self, ctx, opts, args):
+    sample/isolate.
+    """
+    log_lvl = "INFO"
+    if verbose:
+        log_lvl = "DEBUG"
+    elif quiet:
+        log_lvl = "ERROR"
+    logger.remove()
+    logger.add(sys.stderr, level=log_lvl, format=log_fmt)
+    logger.info(f"Welcome to TBpore version {__version__}")
+
+    outdir.mkdir(exist_ok=True, parents=True)
+    if tmp is None:
+        tmp = outdir / TMP_NAME
+    tmp.mkdir(exist_ok=True, parents=True)
+
+    if not inputs:
+        logger.error("No INPUT files given")
+        ctx.exit(2)
+
+    if not name:
+        name = inputs[0].name.split(".")[0]
+        if not name:
+            name = "input"
+
+        logger.debug(f"No sample name found; using '{name}'")
+
+    infile = tmp / f"{name}.fq.gz"
+    if len(inputs) == 1 and inputs[0].is_file():
+        if not inputs[0].suffix == ".gz":
+            with open(inputs[0], "rb") as f_in:
+                with gzip.open(infile, "wb") as f_out:
+                    shutil.copyfileobj(f_in, f_out)
+        else:
+            shutil.copy2(inputs[0], infile)
+
+    else:
+        logger.info("Searching for fastq files...")
+        fq_files = []
+        for p in inputs:
+            if p.is_file():
+                fq_files.append(p)
+            else:
+                fq_files.extend(find_fastq_files(p, recursive))
+        logger.info(f"Found {len(fq_files)} fastq files. Joining them...")
+        concatenate_fastqs(fq_files, infile)
+
+    # todo: download the reference genomes from https://www.nature.com/articles/srep45258#Sec25 and https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0214088
+    # todo: remove contamination
+    # todo: run mykrobe
+    # todo: align decontam to h37rv
+    # todo: bcftools mpileup/call
+    # todo: filter vcf
+    mykrobe = ExternalTool(
+        tool="mykrobe",
+        input=f"-i {infile}",
+        output=f"-o {tmp}/{name}.mykrobe.json",
+        params=f"predict --ont --force -A --debug --format json --min_proportion_expected_depth 0.20 --sample {name} "
+               f"--species tb -t {threads} -m {mem_mb}MB --tmp {tmp} --skeleton_dir {tmp}"
+    )
+
+    subsampled_reads = f"{tmp}/{name}.subsampled.fastq.gz"
+    rasusa = ExternalTool(
+        tool="rasusa",
+        input=f"-i {infile}",
+        output=f"-o {subsampled_reads}",
+        params=f"-c 150 -g 4411532 -s 88"
+    )
+
+    sam_file = f"{tmp}/{name}.subsampled.sam"
+    minimap = ExternalTool(
+        tool="minimap2",
+        input=f"{H37RV_genome} {subsampled_reads}",
+        output=f"-o {sam_file}",
+        params=f"-a -L --sam-hit-only --secondary=no -x map-ont -t {threads}"
+    )
+
+    sorted_sam_file = f"{tmp}/{name}.subsampled.sorted.sam"
+    samtools_sort = ExternalTool(
+        tool="samtools",
+        input=sam_file,
+        output=f"-o {sorted_sam_file}",
+        params=f"sort -@ {threads}"
+    )
+
+    pileup_file = f"{tmp}/{name}.subsampled.pileup.bcf"
```

Comment:
we're not going to keep the pileup file, so you can use VCF, or even stream it directly into `call` if easier

Reply:
I think the `ExternalTool` class does not support streaming, is it ok to leave like this?

Reply:
sure. you can save a tiny bit of time by not using bcf though? can just use vcf. anyway, this is minor

Reply:
Done in https://github.com/mbhall88/tbpore/pull/8/commits/8c3eb8afd323d12afa64c392903a19e762b2f827